File: JAVA/Singleton.java
public class Configuracao {
    private static Configuracao instancia;
    private String valor;

    private Configuracao(String valor) {
        this.valor = valor;
    }

    public static synchronized Configuracao getInstancia(String valor) {
        if (instancia == null) {
            instancia = new Configuracao(valor);
        }
        return instancia;
    }

    public String getValor() {
        return valor;
    }

    public static void main(String[] args) {
        Configuracao config1 = Configuracao.getInstancia("Configuração A");
        Configuracao config2 = Configuracao.getInstancia("Configuração B");

        System.out.println("Config1: " + config1.getValor());
        System.out.println("Config2: " + config2.getValor());
        System.out.println("Config1 e Config2 são a mesma instância: " + (config1 == config2));
    }
}

File: GOLANG/Singleton.go
package main

import (
    "fmt"
    "sync"
)

type Configuracao struct {
    valor string
}

var instancia *Configuracao
var once sync.Once

func ObterConfiguracao(valor string) *Configuracao {
    once.Do(func() {
        instancia = &Configuracao{valor: valor}
    })
    return instancia
}

func main() {
    config1 := ObterConfiguracao("Configuração A")
    config2 := ObterConfiguracao("Configuração B")

    fmt.Println("Config1:", config1.valor)
    fmt.Println("Config2:", config2.valor)
    fmt.Println("Config1 e Config2 são a mesma instância:", config1 == config2)
}
